xdp/src/program.rs

```diff
@@ -0,0 +1,195 @@
+#![allow(clippy::arithmetic_side_effects)]
+
+#[cfg(target_os = "linux")]
+use aya::{programs::Xdp, Ebpf};
+use std::io::{Cursor, Write};
+
+macro_rules! write_fields {
+    ($w:expr, $($x:expr),*) => {
+        $(
+            $w.write_all(&$x.to_le_bytes())?;
+        )*
+    };
+}
+
+const SHT_NULL: u32 = 0;
+const SHT_PROGBITS: u32 = 1;
+const SHT_SYMTAB: u32 = 2;
+const SHT_STRTAB: u32 = 3;
+
+const SHF_ALLOC: u64 = 1 << 1;
+const SHF_EXECINSTR: u64 = 1 << 2;
+
+const STB_GLOBAL: u8 = 1 << 4; // 0x10
+const STT_FUNC: u8 = 2; // 0x02
+
+const XDP_PROG: &[u8] = &[
+    0xb7, 0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x00, // r0 = XDP_PASS
+    0x95, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // exit
+];
+
+const STRTAB: &[u8] = b"\0xdp\0.symtab\0.strtab\0";
+
+pub fn load_xdp_program(if_index: u32) -> Result<Ebpf, Box<dyn std::error::Error>> {
+    let elf = generate_xdp_elf();
+    let mut ebpf = Ebpf::load(&elf).unwrap();
+    let p: &mut Xdp = ebpf.program_mut("xdp").unwrap().try_into().unwrap();
+    p.load()?;
+
+    p.attach_to_if_index(if_index, aya::programs::xdp::XdpFlags::DRV_MODE)?;
+
+    Ok(ebpf)
+}
+
+fn generate_xdp_elf() -> Vec<u8> {
+    let mut buffer = vec![0u8; 4096];
+    let mut cursor = Cursor::new(&mut buffer);
+
+    // start after the header
+    let xdp_off = 64;
+    cursor.set_position(xdp_off);
+    cursor.write_all(XDP_PROG).unwrap();
+    let xdp_size = cursor.position() - xdp_off;
+
+    // write the string table
+    let strtab_off = cursor.position();
+    cursor.write_all(STRTAB).unwrap();
+    let strtab_size = cursor.position() - strtab_off;
+
+    // write the symbol table
+    let symtab_off = align_cursor(&mut cursor, 8);
+    write_symbol(&mut cursor, 0, 0, 0, 0, 0, 0).unwrap();
+    write_symbol(
+        &mut cursor,
+        1, // index
+        0,
+        XDP_PROG.len() as u64,
+        STB_GLOBAL | STT_FUNC,
+        0,
+        1, // section index
+    )
+    .unwrap();
+    let symtab_size = cursor.position() - symtab_off;
+
+    // write the section headers
+    let shdrs_off = align_cursor(&mut cursor, 8);
+    write_section_headers(
+        &mut cursor,
+        xdp_off,
+        xdp_size,
+        strtab_off,
+        strtab_size,
+        symtab_off,
+        symtab_size,
+    )
+    .unwrap();
+
+    // finally go back and write the header
+    const SECTIONS: u16 = 4;
+    const STRTAB_INDEX: u16 = 2;
+    cursor.set_position(0);
+    write_elf_header(&mut cursor, shdrs_off, SECTIONS, STRTAB_INDEX).unwrap();
+
+    buffer
+}
+
+fn align_cursor(cursor: &mut Cursor<&mut Vec<u8>>, alignment: usize) -> u64 {
+    let pos = cursor.position() as usize;
+    let padding = (alignment - (pos % alignment)) % alignment;
+    cursor.set_position((pos + padding) as u64);
+    cursor.position()
+}
+
+fn write_elf_header(
```

Comment:
Would it not be better to use Aya's build approach with cargo xtask rather than literally handrolling assembly?

Reply:
"Would it be better to depend on a new toolchain for 2 assembly instructions?"

No :)

We'll get there, but this is not the time.

Reply:
Alessandro finally got convinced by Dean about superiority of assembly /s

Reply:
😭😭😭

Reply:
Well if the perf difference is so significant then of course asm is justified. Kinda surprising that llvm is so bad at this.

Reply:
maps helpers and anything that calls out of the JIT has overhead. Please let's keep this PR in topic.